### PR TITLE
Fix Two Potential Bugs, with Significant Accuracy Improvement

### PR DESCRIPTION
**Pull request by huanghoujing**
Hi, @jadore801120 , thank you a lot for sharing your transformer implementation.

I find two points that have negative impact on training stability and validation accuracy in your code. And I have made modifications that solve the problem in this pull request.

## Experimental Setting

My experiments are conducted on multi30k-de-en without BPE, with batch size 256, warmup steps 4000 (equivalent to 34 epochs), and total epochs 200 (equivalent to 23400 steps). Other parameters can be found in `train_multi30k_de_en.sh` provided in this commit.

I install Spacy German language model via `pip install de_core_news_sm-2.3.0.tar.gz`, and the English model `conda install -c conda-forge spacy-model-en_core_web_lg`. This is because I encounter connection error while running `python -m spacy download de` and `python -m spacy download en`.

## 1. Learning Rate Magnitude

The learning rate in paper [Attention Is All You Need](https://arxiv.org/abs/1706.03762) is as follows, which corresponds to your code [at this line](https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/fec78a687210851f055f792d45300d27cc60ae41/transformer/Optim.py#L29).

<p align="center">
  
</p>

The learning rate curve can be controlled by two main parameters, warmup steps and multiplication factor.

The curves w.r.t. warmup steps are shown below:

<p align="center">
  
</p>

We can see that, if we want to scale down the maximum learning rate by a half from 7e-4 to 3.5e-4, we have to increase warmup steps from 4000 to 16000. Now consider dataset multi30k-de-en, which has 30000 training samples. When batch size is 256, an epoch contains only 117 steps. Thus 16000 warmup steps is 136 epochs, which would take much longer time to train.

We see that **warmup steps is not a flexible parameter for setting up learning rate**. The better way to use a multiplication factor, as done [in this line of code](https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/fec78a687210851f055f792d45300d27cc60ae41/transformer/Optim.py#L36). I will call it `lr_mul` instead of `init_lr`, since it does not equal to initial learning rate, or highest point of learning rate.

I find that `lr_mul` is hard coded as `2.0` in [`train.py`](https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/fec78a687210851f055f792d45300d27cc60ae41/train.py#L271).

The learning rates, with different multiplication factors, are shown below:

<p align="center">
  
</p>

In the above figure, we observe that, when `lr_mul = 2`, the highest learning rate is 1.4e-3, which is too large as verified in my following experiments. I also find `lr_mul = 1` too large, while `lr_mul = 0.5` is satisfactory. The accuracy and perplexity during training and validation are shown below:
image

From the above figure, we see that **too large a learning rate would drag down the accuracy curve after some period and make it unable to recover**. BTW, this phenomenon was also observed while I used a batch size 768 (with multiple GPUs).

For `lr_mul = 0.5`, the validation accuracy converges within 100 epochs (or 11700 steps). The final performance is
|train accuracy (%) | train ppl | valid accuracy (%) | valid ppl |
|:-:|:-:|:-:|:-:|
|98.253 | 3.94835 | 55.886 | 20.08619 |

I believe these experiments are helpful for solving these issues: #60, #94, #101, #141, #147.

## 2. Scaling Projection or Word Embedding

In paper [Attention Is All You Need](https://arxiv.org/abs/1706.03762), the authors share word embedding layer with output projection layer, but with a scaling factor `sqrt(d_model)` on the embedding.

<p align="center">
  
</p>

In this repository, however, you place the reciprocal scaling factor `1 / sqrt(d_model)` on the projection side, as [in this line of code](https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/fec78a687210851f055f792d45300d27cc60ae41/transformer/Models.py#L158). I find it has been discussed in this issue #138.

I am not sure if this implementation is equivalent to the paper's. However, **I find significant performance gain over the current code, if we implement it on the embedding side.**

I experiment with (1) scaling on the embedding side, (2) scaling on the projection side, (3) no scaling. The learning rate multiplication `lr_mul = 0.5`. The results are as follows:
image

The results show significant difference in convergence speed and validation accuracy. Scaling on the projection side is much slower in convergence, and its validation accuracy is 10.5% lower than on the embedding side (55.886% vs. 66.392%). Maybe scaling the vector before softmax function with `1 / sqrt(d_model) = 1 / sqrt(512) = 0.044` makes softmax difficult to train.

The final performances, with `lr_mul = 0.5` are shown below
| |train accuracy (%) | train ppl | valid accuracy (%) | valid ppl |
|:-:|:-:|:-:|:-:|:-:|
| scaling prj |98.253 | 3.94835 | 55.886 | 20.08619 |
| scaling emb | **99.735** | **3.36672** | **66.392** | 11.73074 |
| no scaling | **99.735** | 3.46923 | 65.478 | **10.35122** |

The slow convergence rate when scaling on the projection side may also provide some hints for this issue: #101.

## Overview of This Pull Request

1. This pull request has made the `lr_mul` parameter configurable.
2. This pull request has made the scaling setting (either on the word embedding side, or on the projection side, or none) configurable.
3. Add seed option for reproducibility of experiments.
4. Save experiment output to directory, making the project directory much cleaner.
5. Optional to use tensorboard to plot curves during training, including perplexity, accuracy, and learning rate, etc.

Once more, I would like to express my appreciation for your sharing the detailed implementation of transformer. I have learned a lot from this code. I hope the pull request would make it clearer for beginners like me to reproduce promising results.

### CODE REVIEW

**Comment by mjkmoynihan** [comment]
Thanks for this PR. It's really informative and helpful!

**Comment by jadore801120** [comment]
@huanghoujing  What a great work! 
I did not notice the two significant parts you experiment on. 
I really appreciate your insights and hard work.
And sorry for the late reply.
